src/sidebar.c

```diff
@@ -409,99 +525,454 @@ static gchar *get_doc_folder(const gchar *path)
+ * caller might want to expand the relevant rows in the tree view */
+static gboolean get_parent_for_file(GtkTreeStore *tree, const gchar *file, GtkTreeIter *parent)
+{
+	gchar *path;
+	GtkTreeIter iter;
+	gint name_diff = 0;
+	gboolean has_parent;
+	GtkTreeModel *model = GTK_TREE_MODEL(tree);
+	TreeForeachData data = {NULL, 0, 0, {0}, TREE_CASE_NONE};
+	gboolean new_row;
+
+	path = g_path_get_dirname(file);
+
+	/* find best opened dir */
+	data.needle = get_doc_folder(path);
+	data.needle_len = strlen(data.needle);
+	name_diff = strlen(path) - data.needle_len;
+	gtk_tree_model_foreach(model, tree_foreach_callback, (gpointer)&data);
+
+	switch (data.best_case)
+	{
+		case TREE_CASE_EQUALS:
+		{
+			*parent = data.best_iter;
+			/* dir already open */
+			new_row = FALSE;
+			break;
+		}
+		case TREE_CASE_CHILD_OF:
+		{
+			/* This dir is longer than existing so just add child */
+			tree_add_new_dir(tree, parent, &data.best_iter, path);
+			new_row = FALSE;
+			break;
+		}
+		case TREE_CASE_PARENT_OF:
+		{
+			/* More complicated logic. This dir should be a parent
+			 * of existing, so reparent existing dir.
+			 */
+			has_parent = gtk_tree_model_iter_parent(model, &iter, &data.best_iter);
+			tree_add_new_dir(tree, parent, has_parent ? &iter : NULL, path);
+			tree_reparent(tree, &data.best_iter, parent);
+			new_row = TRUE;
+			break;
+		}
+		case TREE_CASE_HAVE_SAME_PARENT:
+		{
+			/* Even more complicated logic. Both dirs have same
+			 * parent, so create new parent and reparent them
+			 */
+			GtkTreeIter new_parent;
+			gchar *newpath = g_strndup(path, data.best_len + name_diff);
+
+			has_parent = gtk_tree_model_iter_parent(model, &iter, &data.best_iter);
+			tree_add_new_dir(tree, &new_parent, has_parent ? &iter : NULL, newpath);
+			tree_reparent(tree, &data.best_iter, &new_parent);
+			tree_add_new_dir(tree, parent, &new_parent, path);
+
+			g_free(newpath);
+			new_row = TRUE;
+			break;
+		}
+		default:
+		{
+			tree_add_new_dir(tree, parent, NULL, path);
+			new_row = TRUE;
+			break;
+		}
+	}
+
+	g_free(data.needle);
 	g_free(path);
-	return &parent;
+
+	return new_row;
+}
+
+
+/* Returns true when parent points to a newly added row. */
+static gboolean sidebar_openfiles_add_iter(GtkTreeStore *tree, const gchar *file,
+                                           GtkTreeIter *iter, GtkTreeIter *parent)
+{
+	gboolean new_row;
+	/* get_parent_for_file() might add rows for parent directories */
+	new_row = get_parent_for_file(tree, file, parent);
+	/* insert row for this file */
+	gtk_tree_store_append(tree, iter, parent);
+
+	return new_row;
+}
+
+
+static void expand_iter(GtkTreeIter *iter)
+{
+	GtkTreePath *path;
+
+	path = gtk_tree_model_get_path(GTK_TREE_MODEL(store_openfiles), iter);
+	gtk_tree_view_expand_row(GTK_TREE_VIEW(tv.tree_openfiles), path, FALSE);
+	gtk_tree_path_free(path);
 }
 
 
 /* Also sets doc->priv->iter.
  * This is called recursively in sidebar_openfiles_update_all(). */
+GEANY_EXPORT_SYMBOL
```

Comment:
Why is `GEANY_EXPORT_SYMBOL` necessary here?

Reply:
Again, for the unit test